### PR TITLE
[Proposal] Core Metrics API

### PR DESCRIPTION
**Pull request by dashpole**
Included is a proposal for only collecting what we use in the kubelet, and (eventually) leaving metrics collection for users to third parties.

Issue: #39102

My analysis of what metrics we use and dont use in the kubelet can be [found here](https://docs.google.com/document/d/1Zt1lkb-U7MxI2bg9X2T77MWzhP1FcgppxoplaHr9QVY/edit?usp=sharing).

### CODE REVIEW

**Comment by thockin** [comment]
I don't think I have enough context to review this properly.  Dawn, can you assign?

**Comment by piosz** [comment]
@dashpole I haven't read the whole proposal yet. In the meantime did you have a chance to read [the new monitoring architecture](https://github.com/kubernetes/kubernetes/blob/master/docs/design/monitoring_architecture.md)?

cc @fgrzadkowski

**Comment by dashpole** [comment]
@piosz I will take a look.  Thus far, there appears to be a lot of overlap between the two proposals.

**Comment by dashpole** [comment]
@piosz For the most part, these two proposals look pretty consistent with each other.  This proposal deals with how to collect the core metrics, and which metrics are required, and the other proposal deals with how to create the full pipeline given core metrics.

**Comment by luxas** [comment]
Then I guess we can merge them together?

File: docs/proposals/core-metrics-pipeline.md [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+
+Metrics requirements, based on kubernetes component needs, are as follows:
+ - Kubelet
+  - Node-level capacity and availability metrics for Disk and Memory
+  - Pod-level usage metrics for Disk and Memory
+ - Scheduler
+  - Node-level capacity and availability metrics for Disk, CPU, and Memory
+  - Pod-level usage metrics for Disk, CPU, and Memory
+  - Container-level usage metrics for Disk, CPU, and Memory
+ - Horizontal-Pod-AutoScalar (HPA)
```
Comment (DirectXMan12) [inline docs/proposals/core-metrics-pipeline.md]: Horizontal Pod Auto**scaler**

File: docs/proposals/core-metrics-pipeline.md [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+
+The proposed metrics pipelines would bound the resources that the kubelet consumes, and create a seperation of concerns for kubelet metrics and user metrics.  This proposal also paves the way for simpler third party metrics integration, although that is not covered in this proposal.
+
+### Non Goals
+
+Third Party Metric Provider integration will not be tackled in this proposal. This includes both possibly providing a kubelet metadata (pod-container mappings, volume-pod mappings, etc) API, and a possible standardized metrics APIs for metrics providers to conform to.
+
+Integration with CRI will not be covered in this proposal.  In future proposals, integrating with CRI may provide a better abstraction of information required by the core metrics pipeline to collect metrics.
+
+Methods for pushing these metrics to various kubernetes components will not be covered.  This proposal does not cover any future "metrics-server", but these metrics could easily be pushed to whatever components require them.
```
Comment (DirectXMan12) [inline docs/proposals/core-metrics-pipeline.md]: This *really*  needs to discuss it's place in the monitoring vision in more concrete terms -- if you're referencing the monitoring vision, you could just say here "this proposal does not cover exposing the metrics to components -- that is discussed in more details by the metrics-server"

File: docs/proposals/core-metrics-pipeline.md [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+
+This Core Metrics API be versioned to account for version-skew between kubernetes components.
+
+This proposal purposefully omits many metrics that may eventually become core metrics.  This is by design.  Once metrics are needed for an internal use case, they can be added to the core metrics API.
+
+### Proposed Core Metrics API:
+
+An important difference between the current summary api and the proposed core metrics api is that per-pod stats in the core metrics api contain only usage data, and not capacity-related statistics.  This is more accurate since a pod's resource capacity is really defined by its "requests" and "limits", and it is a better reflection of how the kubelet uses the data.  The kubelet finds which resources are constrained using node-level capacity and availability data, and then chooses which pods to take action on based on the pod's usage of the constrained resource.  If neccessary, capacity for resources a pod consumes can still be correlated with node-level resources using this format of stats.
+
+
```
Comment (DirectXMan12) [inline docs/proposals/core-metrics-pipeline.md]: nit: might want to use fenced code blocks here (like

    ```go
    type FooBar
    ```

so it comes out with syntax highlighting, etc

File: docs/proposals/core-metrics-pipeline.md [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+      Rootfs *DiskUsage `json:"rootfs,omitempty"`  
+      // Stats pertaining to container logs usage of Disk.  
+      // +optional  
+      Logs *DiskUsage `json:"logs,omitempty"`  
+    }  
+
+    // CpuUsage holds statistics about the amount of cpu time consumed  
+    type CpuUsage struct {  
+      // The time at which these stats were updated.  
+      Time metav1.Time `json:"time"`
```
Comment (DirectXMan12) [inline docs/proposals/core-metrics-pipeline.md]: this should probably be `Timestamp` (and elsewhere in this proposal) so it's not confused with CPU time (plus, we use `Timestamp` elsewhere IIRC)

**Review by DirectXMan12 — CHANGES_REQUESTED** [review]
this needs to talk explicitly about how it fits with the monitoring vision (as discussed above), plus some nits and naming questions.

**Comment by DirectXMan12** [comment]
> Then I guess we can merge them together?

I think this should just be trimmed down to not talk about the third-party stuff as much (just reference the monitoring vision).  I think an overhaul of the kubelet part deserves to be it's own sub-proposal, it just needs to be written in terms of the monitoring vision.

**Comment by DirectXMan12** [comment]
P.S. this probably needs to be moved to the kubernetes/community repo.

**Comment by dashpole** [comment]
@DirectXMan12 Ill trim down the vague references to third party integration.  At the time of writing, I was not aware of the other doc which has a much more fully fleshed out story around monitoring architecture.

File: docs/proposals/core-metrics-pipeline.md (hunk 2 above)
Comment (dashpole) [inline docs/proposals/core-metrics-pipeline.md]: done.

File: docs/proposals/core-metrics-pipeline.md (hunk 4 above)
Comment (dashpole) [inline docs/proposals/core-metrics-pipeline.md]: done

**Comment by dashpole** [comment]
@DirectXMan12 Do you think the proposal has the appropriate scope now?  Can you, or someone else who is familiar with the future plans of HPA confirm that the given set of metrics meet the HPA's requirements?

**Comment by luxas** [comment]
@dashpole In any case, all proposals should be directed to kubernetes/community. 
Please reopen this PR there

File: docs/proposals/core-metrics-pipeline.md [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+  - Container-level usage metrics for Disk, CPU, and Memory
+ - Horizontal-Pod-Autoscaler (HPA)
+  - Node-level capacity and availability metrics for CPU and Memory
+  - Pod-level usage metrics for CPU and Memory
+
+More details on how I intend to achieve these high level goals can be found in the Implementation Plan.
+
+In order to continue to provide the full summary API, either the kubelet or a stand-alone version of cAdvisor will need to publish these metrics.
+
+This Core Metrics API be versioned to account for version-skew between kubernetes components.
```
Comment (DirectXMan12) [inline docs/proposals/core-metrics-pipeline.md]: > This Core Metrics API **will** be versioned

**Comment by DirectXMan12** [comment]
scope looks better.  please tag me once you open the new PR

**Comment by luxas** [comment]
New PR up, great!

**Pull request closed without merging**